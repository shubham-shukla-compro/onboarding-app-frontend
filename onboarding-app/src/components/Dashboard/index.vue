<template>
  <div class="all-tasks-container">
    <h1 class="title">{{ title }}</h1>
    <div class="all-tasks-section">
      <router-link to="/add-task">
        <button class="all-tasks-section-btn">
          <i class="fas fa-plus"></i>
          Add Task
        </button>
      </router-link>
      <table class="all-tasks-table" border="1">
        <thead>
          <tr>
            <th>#</th>
            <th>Module</th>
            <th>Contents</th>
            <th>Duration (in days)</th>
            <th>Edit/Delete</th>
          </tr>
        </thead>
        <tbody>
          <tr v-for="task in tasksData" :key="task.id" v-show="!task.finished">
            <td>
              <input
                type="checkbox"
                :value="task"
                v-model="checkedTasks"
                @click="onClick(task)"
              />
            </td>
            <td>{{ task.module }}</td>
            <td>{{ task.contents }}</td>
            <td>{{ task.duration }}</td>
            <td>
              <router-link
                :to="{ path: '/edit-task/' + task.id }"
                class="all-task-edit-btn"
              >
                <i class="fas fa-edit"></i>
              </router-link>
              <button class="all-task-delete-btn" @click="onDelete(task.id)">
                <i class="fas fa-trash"></i>
              </button>
            </td>
          </tr>
          <!-- completed -->
          <tr v-show="checkedTasks.length">
            <th colspan="5">Completed Tasks</th>
          </tr>
          <tr v-for="task in checkedTasks" :key="task.id">
            <td>
              <input
                type="checkbox"
                :value="task"
                v-model="checkedTasks"
                @click="onClick(task)"
              />
            </td>
            <td>{{ task.module }}</td>
            <td>{{ task.contents }}</td>
            <td>{{ task.duration }}</td>
            <td>
              <router-link
                :to="{ path: '/edit-task/' + task.id }"
                class="all-task-edit-btn"
              >
                <i class="fas fa-edit"></i>
              </router-link>
              <button class="all-task-delete-btn" @click="onDelete(task.id)">
                <i class="fas fa-trash"></i>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</template>

<script>
import axios from 'axios';

export default {
  data() {
    return {
      title: 'OnBoarding Tasks',
      tasksData: [],
      checkedTasks: [],
    };
  },
  methods: {
    onClick(task) {
      task.finished = !task.finished;
      this.tasksData.map((tdata) => {
        if (tdata.id === task.id) {
          tdata = task;
        }
        return tdata;
      });
      this.updateData(task);
      return this.tasksData;
    },
    onDelete(index) {
      this.checkedTasks = this.checkedTasks.filter((ctask) => {
        return ctask.id !== index;
      });
      this.tasksData = this.tasksData.filter((tdata) => {
        return tdata.id !== index;
      });
      this.deleteData(index);
    },

    async getAllData() {
      try {
        const res = await axios.get('http://localhost:8000/tasks');
        if (res.status === 200) {
          const localData = res.data;
          if (localData) {
            this.tasksData = localData;
            localData.map((ldata) => {
              if (ldata.finished) {
                this.checkedTasks.push(ldata);
              }
              return this.checkedTasks;
            });
          }
        }
        // console.log(res.data);
        return this.tasksData;
      } catch (err) {
        console.error(err.message);
      }
    },

    async updateData(task) {
      try {
        const config = {
          headers: {
            'Content-Type': 'application/json',
          },
        };
        const body = JSON.stringify(task);
        const res = await axios.put(
          `http://localhost:8000/tasks/${task.id}`,
          body,
          config
        );
        if (res.status !== 200) {
          console.log(res);
        }
      } catch (err) {
        console.log(err.message);
      }
    },

    async deleteData(id) {
      try {
        const res = await axios.delete(`http://localhost:8000/tasks/${id}`);
        if (res.status === 200) {
          // this.getAllData();
        }
      } catch (err) {
        console.error(err.message);
      }
    },
  },

  created() {
    this.getAllData();
  },
};
</script>

<style scoped src="./styles.css"></style>
