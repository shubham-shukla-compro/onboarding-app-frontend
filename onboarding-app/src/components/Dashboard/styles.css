.all-tasks-container {
  margin: 10px auto;
  padding: 20px;
  max-width: 900px;
}

.all-tasks-section {
  background-color: var(--white-color);
  width: 100%;
  height: 100%;
  padding: 10px;
}

.all-tasks-section-btn {
  display: block;
  padding: 14px 20px;
  margin: 35px auto;
  cursor: pointer;
  border: none;
  border-radius: 5px;
  font-size: 1rem;
  float: right;
}

.title {
  text-align: center;
}

.all-tasks-table {
  width: 100%;
  border-collapse: collapse;
  font-size: 1.2rem;
  background-color: var(--white-color);
}

.all-tasks-table tr th {
  color: var(--white-color);
  background-color: #2f2f2f;
  border-bottom: solid 2px #d8d8d8;
  position: sticky;
  top: 0;
  white-space: nowrap;
  text-align: center;
  padding: 15px;
  box-sizing: border-box;
}

.all-tasks-table tr td {
  border: solid 1px #d8d8d8;
  text-align: center;
  padding: 15px;
}

.all-tasks-table tbody tr {
  transition: background-color 150ms ease-out;
}
.all-tasks-table tbody tr:nth-child(2n) {
  background-color: whitesmoke;
}
.all-tasks-table tbody tr:hover {
  background-color: #ebebeb;
}
