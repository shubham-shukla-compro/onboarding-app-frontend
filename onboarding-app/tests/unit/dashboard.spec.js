import { shallowMount } from '@vue/test-utils';
import Dashboard from '@/components/Dashboard';

describe('Dashboard.vue', () => {
  let wrapper;
  const msg = 'OnBoarding Tasks';

  describe('#test', () => {
    beforeEach(() => {
      wrapper = shallowMount(Dashboard, {
        propsData: { msg },
        stubs: ['router-link'],
      });
    });

    it('renders props.msg when passed', () => {
      const title = wrapper.find('.title');

      expect(title.text()).toMatch(msg);
      wrapper.destroy();
    });

    it('should be a vue instance', () => {
      expect(wrapper).toBeTruthy();
      wrapper.destroy();
    });

    it('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
      wrapper.destroy();
    });
  });

  describe('#Delete Task', () => {
    const tasksData = [
      {
        id: 1,
        module: 'Test Module',
        contents: 'Test Contents',
        duration: 2,
        finished: false,
      },
    ];
    beforeEach(() => {
      wrapper = shallowMount(Dashboard, {
        data() {
          return {
            tasksData,
          };
        },
        stubs: ['router-link'],
      });
    });

    it('should delete a task', async () => {
      const onDelete = jest.fn();
      wrapper.setMethods({
        onDelete,
      });
      const deleteBtn = wrapper.find('.all-task-delete-btn');
      await deleteBtn.trigger('click');
      expect(onDelete).toHaveBeenCalled();
      wrapper.destroy();
    });
  });
});
