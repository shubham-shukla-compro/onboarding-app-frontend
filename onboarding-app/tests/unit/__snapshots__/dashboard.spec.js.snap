// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard.vue #test snapshot 1`] = `
<div class="all-tasks-container" msg="OnBoarding Tasks">
  <h1 class="title">OnBoarding Tasks</h1>
  <div class="all-tasks-section">
    <router-link-stub to="/add-task"><button class="all-tasks-section-btn"><i class="fas fa-plus"></i>
        Add Task
      </button></router-link-stub>
    <table border="1" class="all-tasks-table">
      <thead>
        <tr>
          <th>#</th>
          <th>Module</th>
          <th>Contents</th>
          <th>Duration (in days)</th>
          <th>Edit/Delete</th>
        </tr>
      </thead>
      <tbody>
        <tr style="display: none;">
          <th colspan="5">Completed Tasks</th>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
